### PR TITLE
Add ServiceMonitor to Helm

### LINKED ISSUE #194: Create ServiceMonitor resource in Helm
To accomplish https://github.com/aquasecurity/trivy-operator/issues/193 we need to create a ServiceMonitor to tell Prometheus to scrape trivy-operator metrics on `/metrics` endpoint in a more elegant way. So we can have more control on the metrics endpoint.

### PR DESCRIPTION
**Pull request by Dentrax**
Signed-off-by: Furkan <furkan.turkal@trendyol.com>

## Description

Screen Shot 2022-06-27 at 14 28 11

## Related issues
- Close #194

Remove this section if you don't have related PRs.

## Checklist
- [X] I've read the [guidelines for contributing](https://github.com/aquasecurity/trivy-operator/blob/main/CONTRIBUTING.md) to this repository.
- [ ] I've added tests that prove my fix is effective or that my feature works.
- [ ] I've updated the [documentation](https://github.com/aquasecurity/trivy-operator/tree/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

PTAL @eminaktas

### CODE REVIEW

File: deploy/helm/templates/servicemonitor.yaml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+{{- if eq .Values.serviceMonitor.enabled true }}
+apiVersion: monitoring.coreos.com/v1
+kind: ServiceMonitor
+metadata:
+  name: {{ include "trivy-operator.fullname" . }}
+  namespace: {{ .Release.Namespace }}
```
Comment (eminaktas) [inline deploy/helm/templates/servicemonitor.yaml:6]:
```suggestion
  namespace: {{ .Values.serviceMonitor.namespace | default .Release.Namespace }}
```

File: deploy/helm/templates/servicemonitor.yaml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  labels:
+    {{- include "trivy-operator.labels" . | nindent 4 }}
+spec:
+  namespaceSelector:
+    matchNames:
+    - {{ .Release.Namespace }}
```
Comment (eminaktas) [inline deploy/helm/templates/servicemonitor.yaml:12]:
```suggestion
  {{- if .Values.serviceMonitor.namespace }}
  namespaceSelector:
    matchNames:
    - {{ .Release.Namespace }}
  {{- end }}
```

We don't need to specify until the service is in another namespace.

File: deploy/helm/values.yaml [hunk 3]
```diff
@@ -69,11 +69,12 @@ image:
...
-  type: ClusterIP
   metricsPort: 80
-  annotations:
-    prometheus.io/scrape: "true"
-    prometheus.io/path: /metrics
+
+serviceMonitor:
+  enabled: true
+  interval: ""
+  # honorLabels: true
```
Comment (eminaktas) [inline deploy/helm/values.yaml:78]:
```suggestion
  # The namespace where Prometheus expects to find service monitors.
  # namespace: ""

```

**Comment by chen-keinan** [comment]
@Dentrax Thanks for the contribution, could you pls run the ./hack/update-static.yaml.sh in order to update the static files

**Comment by chen-keinan** [comment]
> @Dentrax Thanks for the contribution, could you pls run the ./hack/update-static.yaml.sh in order to update the static files and push the changes

@Dentrax Pls update the ./hack/update-static.yaml.sh to include the service.yaml and servicemonitor.yaml

**Comment by Dentrax** [comment]
So it better to set it `false` by default? I have also removed the `prometheus/scrape` annotations in favor of ServiceMonitor. Should I bring back those annotations?

**Comment by erikgb** [comment]
> So it better to set it `false` by default? I have also removed the `prometheus/scrape` annotations in favor of ServiceMonitor. Should I bring back those annotations?

Well, I wouldn't worry too much about compatibility, as trivy-operator is quite fresh and has not reached GA yet. I would expect most users interested in metrics to have the prom-operator installed. So I'll wote for replacing the annotations with ServiceMonitor. But as users might experience installation errors without the CRD present, I'll prefer if this is somehow mentioned in the (installation) docs. WDYT @josedonizetti @chen-keinan ?

**Comment by chen-keinan** [comment]
> > So it better to set it `false` by default? I have also removed the `prometheus/scrape` annotations in favor of ServiceMonitor. Should I bring back those annotations?
> 
> Well, I wouldn't worry too much about compatibility, as trivy-operator is quite fresh and has not reached GA yet. I would expect most users interested in metrics to have the prom-operator installed. So I'll wote for replacing the annotations with ServiceMonitor. But as users might experience installation errors without the CRD present, I'll prefer if this is somehow mentioned in the (installation) docs. WDYT @josedonizetti @chen-keinan ?

it should be definitely documented , we will be adding section for `Prometheus Integration` in our docs and include the pre-requisite for CRDs installation.
Still I would set `serviceMonitor.enabled : false` by default as I do not thinks that the absence of `ServiceMonitor` crd should fail the helm install for trivy-operator.

**Comment by chen-keinan** [comment]
> 

I do not see the `./hack/update-static.yaml.sh` file in changes.  after changing the file it need to be executed and committing the changed static files as well

File: deploy/helm/values.yaml [hunk 4]
```diff
@@ -69,11 +69,14 @@ image:
 # service only expose a metrics endpoint for prometheus to scrape,
 # trivy-operator does not have a user interface.
 service:
-  type: ClusterIP
   metricsPort: 80
-  annotations:
-    prometheus.io/scrape: "true"
-    prometheus.io/path: /metrics
+
+serviceMonitor:
```
Comment (josedonizetti) [inline deploy/helm/values.yaml:74]:
```suggestion
# Prometheus ServiceMonitor configuration
# See: https://github.com/prometheus-operator/prometheus-operator
serviceMonitor:
```

**Review by josedonizetti — CHANGES_REQUESTED** [review]
@Dentrax  Thank you for the contribution. Let's change the monitor to be disabled by default. Also, take a look on @chen-keinan comments about `update-static.sh`

**Comment by Dentrax** [comment]
Thanks! Changes have been made according to your reviews. I didn't touch official docs, I think it'd be most suitable for another PR. (Since there is no Prometheus section requires a bit of initial work)

File: hack/update-static.yaml.sh [hunk 5]
```diff
@@ -18,6 +18,8 @@ cp $HELM_TMPDIR/trivy-operator-helm-template/trivy-operator/templates/rbac.yaml
 cp $HELM_TMPDIR/trivy-operator-helm-template/trivy-operator/templates/config.yaml $STATIC_DIR/03-trivy-operator.config.yaml
 cp $HELM_TMPDIR/trivy-operator-helm-template/trivy-operator/templates/policies.yaml $STATIC_DIR/04-trivy-operator.policies.yaml
 cp $HELM_TMPDIR/trivy-operator-helm-template/trivy-operator/templates/deployment.yaml $STATIC_DIR/05-trivy-operator.deployment.yaml
+cp $HELM_TMPDIR/trivy-operator-helm-template/trivy-operator/templates/service.yaml $STATIC_DIR/05-trivy-operator.service.yaml
```
Comment (josedonizetti) [inline hack/update-static.yaml.sh:21]:
```suggestion
cp $HELM_TMPDIR/trivy-operator-helm-template/trivy-operator/templates/service.yaml $STATIC_DIR/06-trivy-operator.service.yaml
```

File: hack/update-static.yaml.sh [hunk 6]
```diff
@@ -18,6 +18,8 @@ cp $HELM_TMPDIR/trivy-operator-helm-template/trivy-operator/templates/rbac.yaml
...
+cp $HELM_TMPDIR/trivy-operator-helm-template/trivy-operator/templates/servicemonitor.yaml $STATIC_DIR/05-trivy-operator.servicemonitor.yaml
```
Comment (josedonizetti) [inline hack/update-static.yaml.sh:22]:
```suggestion
cp $HELM_TMPDIR/trivy-operator-helm-template/trivy-operator/templates/servicemonitor.yaml $STATIC_DIR/07-trivy-operator.servicemonitor.yaml
```

File: hack/update-static.yaml.sh [hunk 7]
```diff
@@ -18,6 +18,8 @@ cp $HELM_TMPDIR/trivy-operator-helm-template/trivy-operator/templates/rbac.yaml
...
 
 cat $CRD_DIR/vulnerabilityreports.crd.yaml \
```
Comment (josedonizetti) [inline hack/update-static.yaml.sh:24]: We need to add the service here as well, to aggregate it on `trivy-operator.yaml` (after adding it we need to run `./hack/update-static.sh`

**Review by josedonizetti — COMMENTED** [review]
@Detranx Thinking about it now, maybe when the service monitor is disabled we should have the service use ClusterIP + annotations like it were before. WDYT?

@chen-keinan @erikgb thoughts?

**Review by chen-keinan — CHANGES_REQUESTED** [review]

File: hack/update-static.yaml.sh [hunk 8]
```diff
@@ -27,4 +29,6 @@ cat $CRD_DIR/vulnerabilityreports.crd.yaml \
   $STATIC_DIR/02-trivy-operator.rbac.yaml \
   $STATIC_DIR/03-trivy-operator.config.yaml \
   $STATIC_DIR/04-trivy-operator.policies.yaml \
-  $STATIC_DIR/05-trivy-operator.deployment.yaml > $STATIC_DIR/trivy-operator.yaml
+  $STATIC_DIR/05-trivy-operator.deployment.yaml \
+  $STATIC_DIR/05-trivy-operator.service.yaml \
+  $STATIC_DIR/05-trivy-operator.servicemonitor.yaml \ > $STATIC_DIR/trivy-operator.yaml
```
Comment (josedonizetti) [inline hack/update-static.yaml.sh:34]: @Dentrax Thank you for the patience here. To wrap this up, two things only:

1 - fix the file numbers as per my review above
2 - dont' add `servicemonitor.yaml` here, or else static deploy will depend on prometheus-operator

Thank you!

Comment (Dentrax) [reply hack/update-static.yaml.sh]: Oh, thanks! Also sorry I don't get the idea of this `sh` file, but I think It'd be better to handle these stuff in automated way such as a `make` file or smth.

Done!

**Comment by Dentrax** [comment]
> maybe when the service monitor is disabled we should have the service use ClusterIP + annotations like it were before

Makes sense, but also @erikgb have mentioned:

> I'll wote for replacing the annotations with ServiceMonitor.

But keeping these annotations as is when `serviceMonitor.enabled: false` sounds more compatible for me.

**Comment by chen-keinan** [comment]
@Dentrax thanks you for the effort. The code generation need to be called again after every change in order to be consist.
- `05-trivy-operator.service.yaml` need to be deleted
- `./hack/./update-static.yaml.sh` need to be executed , as result of it it will 
    - amend `deploy/static/trivy-operator.yaml` 
    - add `deploy/static/06-trivy-operator.service.yaml`

**Comment by chen-keinan** [comment]
@Dentrax let me know if you want me to do for you that last changes for generating the static files

**Comment by Dentrax** [comment]
> @Dentrax let me know if you want me to do for you that last changes for generating the static files

Hey @chen-keinan, just updated the files and committed according to your instructions. Please feel free to commit my branch if i did wrong, `update-static.yaml.sh` script slightly confused me. 🙈

**Comment by josedonizetti** [comment]
@Dentrax changes look good, needs rebase though to fix the version.

**Comment by chen-keinan** [comment]
@Dentrax LGTM 🚀  thanks for the contribution

**Comment by Dentrax** [comment]
Thanks, rebased. Can you please check again?

**Review by josedonizetti — APPROVED** [review]

**Comment by chen-keinan** [comment]
LGTM 🚀

**Pull request merged**